Pileup and calibration DA for PHOS and CPV

peressounko opened the pull request:
- fixed issue with same bunch pileup in PHOS and CPV

- adds option raw->cluster reconstruction for CPV

- adds calibration algorithms for calibration for PHOS and CPV and skeleton for relative E calibration in PHOS

- fix argument range in calibration objects

- PHOS mis-alignment matrices added

shahor02 commented:
Hi @peressounko 

The CI-s are failing with:
````
+ value='[20:10:24][ERROR] Alignable entry CPV/Module3 NOT set'
+ '[' -n '[20:10:24][ERROR] Alignable entry CPV/Module3 NOT set' ']'
````

Would be better to not mix so many unrelated commits in a single PR...

Cheers,
 Ruben

sawenzel commented:
I confirm that the in-bunch pileup test now passes for PHS/CPV digits. But I would also prefer one feature = one PR. The geometry fixes probably have nothing to do with pileup. Just as a wish for future developments.

peressounko commented on Detectors/PHOS/calib/PHOSCalibWorkflow/include/PHOSCalibWorkflow/PHOSCalibCollector.h:
```diff
@@ -0,0 +1,195 @@
+ public:
+  PHOSCalibCollector() = default;
+  PHOSCalibCollector(short mode) : mMode(mode) {}
+
+  ~PHOSCalibCollector() final = default;
+
```
Dear @shahor02 ,
what code checked suggests here? Initial version with ~PHOSCalibCollector() = default; checker criticized and suggested adding final/override. This version also causes warning. I think, this should be final class and I do not want to write descructor explicitely. What should I write here? Thanks, Dmitri

shahor02 replied:
@peressounko the LLVM produces a warning (which we interpret as an error) if the class having a virtual destructor and not flagged as ``final`` declares its destructor as ``final``. Use ``override = default`` instead.

peressounko commented:
Dear @sawenzel @shahor02, initially I prepared PR with calibration algorithms. But then I added also a fix of pileup issue which looks more urgent. Can we go with this PR (when clear code checker warnings) or should I split it into two? Best regards, Dmitri

shahor02 commented:
@peressounko fine for me to merge once the CIs are passed, but would be good if you could squash commits, grouping them logically consistent way (and getting rid of fixups like clanged).

Commit:
PHOS,CPV calibration algorithms added

Commit:
Pileup simulation/reconstruction fixed

peressounko commented:
build/O2/o2-dataflow has errors in QC checks, probably not related to this PR

shahor02 commented on Detectors/CPV/calib/CPVCalibWorkflow/include/CPVCalibWorkflow/CPVBadMapCalibDevice.h:
```diff
@@ -0,0 +1,64 @@
+  std::string mPath{"./"};                                  ///< path and name of file with collected histograms
+  std::unique_ptr<BadChannelMap> mBadMap;                   /// Final calibration object
+  std::array<char, o2::cpv::Geometry::kNCHANNELS> mMapDiff; /// difference between new and old map
+  ClassDefNV(CPVBadMapCalibDevice, 1);
```
Please remove this, there is no point creating a dictionary for DPL Tasks (and if there was, since the Task is a virtual class, the ClassDefOverride(..) should have been used).

peressounko replied:
removed

shahor02 commented on Detectors/CPV/calib/CPVCalibWorkflow/include/CPVCalibWorkflow/CPVGainCalibDevice.h:
```diff
@@ -0,0 +1,66 @@
+  std::unique_ptr<CalibParams> mCalibParams;                   /// Final calibration object
+  std::unique_ptr<TH2F> mMean;                                 /// Mean values in High Gain channels
+  std::array<short, o2::cpv::Geometry::kNCHANNELS> mGainRatio; /// Gain variation wrt previous map
+  ClassDefNV(CPVGainCalibDevice, 1);
```
Same comment on ClassDefNV for Task

peressounko replied:
removed

shahor02 commented on Detectors/PHOS/calib/PHOSCalibWorkflow/include/PHOSCalibWorkflow/PHOSHGLGRatioCalibDevice.h:
```diff
@@ -0,0 +1,82 @@
+  std::unique_ptr<CalibParams> mCalibObject;                  /// Final calibration object
+  std::unique_ptr<TH2F> mhRatio;                              /// Histogram with ratios
+  std::array<float, o2::phos::Mapping::NCHANNELS> mRatioDiff; /// Ratio variation wrt previous map
+  ClassDefNV(PHOSHGLGRatioCalibDevice, 1);
```
Idem

peressounko replied:
removed

shahor02 commented on Detectors/PHOS/calib/PHOSCalibWorkflow/include/PHOSCalibWorkflow/PHOSPedestalCalibDevice.h:
```diff
@@ -0,0 +1,72 @@
+  std::unique_ptr<TH2F> mRMSLG;                               /// RMS of values in Low Gain channels
+  std::array<short, o2::phos::Mapping::NCHANNELS> mPedHGDiff; /// Pedestal variation wrt previous map
+  std::array<short, o2::phos::Mapping::NCHANNELS> mPedLGDiff; /// Pedestal variation wrt previous map
+  ClassDefNV(PHOSPedestalCalibDevice, 1);
```
Idem

peressounko replied:
removed

shahor02 commented on Detectors/CPV/calib/CPVCalibWorkflow/src/CPVGainCalibDevice.cxx:
```diff
@@ -0,0 +1,231 @@
+  mMean = std::unique_ptr<TH2F>(new TH2F("Gains", "Signals per channel", o2::cpv::Geometry::kNCHANNELS, 0.5, o2::cpv::Geometry::kNCHANNELS + 0.5, 1024, 0., 4096.));
+
+  std::string filename = mPath + "CPVGains.root";
+  TFile filein(filename.data(), "READ");
+  if (filein.IsOpen()) {
+    mMean->Add(static_cast<TH2F*>(filein.Get("Gains")));
+    filein.Close();
+  }
```
You are reading the file into the histo, filling the latter and save it again, is this in order to accumulate over multiple runs? In this way you don't have a control over what you get from the disk, if such an accumulation is needed, it should be done either via CCDB, or you via filling containers which are then merged and processed.

peressounko replied:
Yes, this calibration may require scan of several runs. These are transient data with relatively large size, probably CCDB is not good place to store them. Files with histos can be used as "containers" and merged if necessary with root tools. I plan to provide some configuration with paths to files etc. Or we have this functionality implemented in framework?

shahor02 replied:
Concerning CCDB, I meant to put there not transient histos but to incrementally cumulate the calibration, if the algorithm allows this. How are you going to control the period over which the integration goes? 
Also, do you really need to update the CCDB in the end of the workflow, if in the next session you still keep updating this file? Are these objects used for the sync. workflow?
Eventually, the containers will be handled by the framework. Anyway, the changes can be itnroduced later.

peressounko replied:
For gain-like calibrations there is no much sense in updating with intermediate statistics, so only final objects should be sent. I plan to check mean statistics per channel and finalize calibration once threshold is set. This will probably require manual intervention, but this should be rare procedure ~ 1 time per year. Normally CCDB will not be updated per processing at EOS, only temporary objects will be stored to files. Only when statistics will be reached calibration will be send to CCDB. Output to QC will be send on each end of workflow and experts will be able to look at status. 
For PHOS some versions of gain will be used in reconstruction but final calibration will be applied on analysis level. For CPV gain calibration will be necessary for clustering. But it should be produced on commissioning stage.

shahor02 commented on Detectors/CPV/calib/CPVCalibWorkflow/src/CPVPedestalCalibDevice.cxx:
```diff
@@ -0,0 +1,253 @@
+    mUpdateCCDB = true;
+    return;
+  }
+  // //TODO:
+  // //Get current map
+  // int nChanged=0;
+  // for(short i=o2::cpv::Geometry::kNCHANNELS; --i;){
+  //   short dp=mPedestals.getPedestal(i)-oldPed.getPedestal(i);
+  //   mPedDiff[i]=dp ;
+  //   if(abs(dp)>1){ //not a fluctuation
+  //     nChanged++;
+  //   }
+  // }
+  // if(nChanged>kMinorChange){ //serious change, do not update CCDB automatically, use "force" option to overwrite
+  //   mUpdateCCDB=false;
+  // }
+  // else{
+  //   mUpdateCCDB=true;
+  // }
+}
```
Are those commented methods to be refined later? If not, better to suppress the dead code.

peressounko replied:
Yes, I will implement them together with sending/reading to CCDB.

shahor02 commented on Detectors/CPV/calib/CPVCalibWorkflow/src/CPVPedestalCalibDevice.cxx:
```diff
@@ -0,0 +1,253 @@
+                                          o2::framework::Options{}};
+}
+
+// //============================================================
+// DataProcessorSpec getTOFChannelCalibDeviceSpec(bool useCCDB){
+//   using device = o2::calibration::TOFChannelCalibDevice;
+//   using clbUtils = o2::calibration::Utils;
+
```
this is just commented TOF code, I guess can be eliminated

peressounko replied:
removed

peressounko replied:
Dear @shahor02, @chiarazampolli, we will have two kinds of calibrations: pedestals (CPV) and pedestals, HG/LG ratio (PHOS) should process data from dedicated runs. Full run, usually relatively short ~few minutes should be processed. Another kind of calibrations: gains (CPV), time and gain (PHOS) will require several runs and even fills. In some discussions before we agreed that transient data should be stored on disk of DPL computers and processed when sufficient statistics will be collected. I tried to use corresponding lifetimes of devices. Do I miss something?

shahor02 replied:
We were speaking about the calibration info which is on the EPN accumulated (merged) in the containers, processed offline then deleted. These EPN containers, since there will be 250 of them,  will need to be aggregated, then the calibration can created. 
In case of your calib. devices, I am not sure: since you are using raw data, I guess they will be running on the FLP, since as far as understand, you will run the cell fits on FLPs and will not ship the raw data to EPN. I am not sure there we can accumulate data there (was never discussed), so, we may need to produce on FLP some intermediate data from raw, and send to EPN for accumulation.

peressounko replied:
Calibrations like Pedestals and HG/LG ratio do use raw data. Their output histos (in addition to CCDB objects) will be used for bad map and should be shipped to EPN. These are light-weight devices and should not consume a lot of CPU. Even if they will interfere with data taking, I think this is not a problem if part of pedestal or LED runs will be lost. Another option: split device to standard raw decoding part on FLP and analysis part on EPN. 
PHOS time and gain device will use clusters and reside on EPN. CPV gain device can use raw data as now or digits if we decide to move it to EPN. We will discuss this with CPV experts.

shahor02 reviewed (changes requested):
Hi @peressounko,
 
Please see some comments below, the ClassDefs for DPL tasks should be removed.
I did not test the code, but as far as I understood, @sawenzel is happy with the pile-up performance, so after fixing this for me it is fine to be merged.

But I have general question about the calibration devices (if needed, can be reimplemented later): I see that you typically create some histos in the init, fill them through all TFs, then send the calibrated object to CDB-populator. In this way the granularity of your objects will be run, fill or even few of them, depending on how often the workflow started / stopped. 
Is this what you want? We have user-defined time-window calibration framework, you copy-pasted some TOF code from @chiarazampolli , why not relying fully on this framework, which gives you control on the coverage of the calib. objects?

Cheers,
 Ruben
